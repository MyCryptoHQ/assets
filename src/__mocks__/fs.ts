import { basename } from 'path';

export const promises = {
  mkdir: jest.fn(),

  readFile: jest.fn(
    async (path: string, encoding: string): Promise<string> => {
      const file = basename(path);

      switch (file) {
        case 'eth.json':
          return `
            [
              {
                "address": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
                "symbol": "GNT",
                "decimal": 18,
                "name": "Golem",
                "uuid": "59a6f871-34bb-5914-86cc-b5f1307bb3b9"
              },
              {
                "address": "0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2",
                "symbol": "MKR",
                "decimal": 18,
                "name": "MakerDAO",
                "uuid": "d3230312-cacb-52e6-a849-8bbf6e8142dd"
              }
            ]
          `;
        case 'assets.json':
          return `
          {
            "356a192b-7913-504c-9457-4d18c28d46e6": {
              "coinCapId": "ethereum",
              "coinGeckoId": "ethereum",
              "cryptoCompareId": "ETH",
              "cryptoCurrencyIconName": "eth",
              "invalid": "field"
            },
            "59a6f871-34bb-5914-86cc-b5f1307bb3b9": {
              "coinCapId": "golem-network-tokens",
              "coinGeckoId": "golem",
              "cryptoCompareId": "GNT",
              "cryptoCurrencyIconName": "gnt"
            },
            "d3230312-cacb-52e6-a849-8bbf6e8142dd": {
              "coinCapId": "maker",
              "coinGeckoId": "maker",
              "cryptoCompareId": "MKR",
              "cryptoCurrencyIconName": "mkr"
            }
          }
        `;
      }

      throw new Error('File not found');
    }
  ),

  writeFile: jest.fn()
};
