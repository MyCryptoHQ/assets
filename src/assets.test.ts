import { promises as fs } from 'fs';
import execa from 'execa';
import {
  fetchEthereumTokens,
  getAssets,
  getIds,
  getParsedAssets,
  parseTokensJson,
  writeToDisk
} from './assets';

jest.mock('fs');
jest.mock('execa');
jest.mock('./services', () => ({
  matchCoinCapId: jest.fn(async () => 'golem-network-tokens'),
  matchCoinGeckoId: jest.fn(async () => 'golem'),
  matchCryptoCompareId: jest.fn(async () => 'GNT'),
  matchCryptoCurrencyIcon: jest.fn(),
  matchDexAgId: jest.fn(async () => 'GNT')
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('getIds', () => {
  it('gets all (available) service IDs for an asset', async () => {
    const asset = {
      address: '0xa74476443119A942dE498590Fe1f2454d7D4aC0d',
      decimal: '18',
      name: 'Golem',
      symbol: 'GNT',
      uuid: '59a6f871-34bb-5914-86cc-b5f1307bb3b9'
    };

    await expect(getIds(asset, null)).resolves.toMatchSnapshot();
  });

  it('uses existing parsed values if available', async () => {
    const asset = {
      address: '0xa74476443119A942dE498590Fe1f2454d7D4aC0d',
      decimal: '18',
      name: 'Golem',
      symbol: 'GNT',
      uuid: '59a6f871-34bb-5914-86cc-b5f1307bb3b9'
    };

    const parsedAsset = {
      cryptoCurrencyIconName: 'gnt'
    };

    const ids = await getIds(asset, parsedAsset);
    expect(ids).toHaveProperty('cryptoCurrencyIconName', 'gnt');
  });

  it('overwrites existing values if they changed', async () => {
    const asset = {
      address: '0xa74476443119A942dE498590Fe1f2454d7D4aC0d',
      decimal: '18',
      name: 'Golem',
      symbol: 'GNT',
      uuid: '59a6f871-34bb-5914-86cc-b5f1307bb3b9'
    };

    const parsedAsset = {
      coinCapId: 'golem',
      cryptoCurrencyIconName: 'gnt'
    };

    const ids = await getIds(asset, parsedAsset);
    expect(ids).toHaveProperty('coinCapId', 'golem-network-tokens');
  });
});

describe('parseTokensJson', () => {
  it('parses a JSON string to a tokens array', async () => {
    const json = `
      [
        {
          "address": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
          "symbol": "GNT",
          "decimal": 18,
          "name": "Golem",
          "uuid": "59a6f871-34bb-5914-86cc-b5f1307bb3b9"
        },
        {
          "address": "0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2",
          "symbol": "MKR",
          "decimal": 18,
          "name": "MakerDAO",
          "uuid": "d3230312-cacb-52e6-a849-8bbf6e8142dd"
        }
      ]
    `;

    await expect(parseTokensJson(json)).resolves.toMatchSnapshot();
  });

  it('throws on invalid tokens', async () => {
    const json = `
      [
        {
          "address": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
          "symbol": "GNT",
          "decimal": 18,
          "name": "Golem",
          "uuid": "59a6f871-34bb-5914-86cc-b5f1307bb3b9"
        },
        {
          "address": "0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2",
          "symbol": "MKR",
          "decimal": -1,
          "name": "MakerDAO",
          "uuid": "d3230312-cacb-52e6-a849-8bbf6e8142dd"
        }
      ]
    `;

    await expect(parseTokensJson(json)).rejects.toThrow();
  });
});

describe('fetchEthereumTokens', () => {
  it('fetches Ethereum tokens from the disk', async () => {
    await expect(fetchEthereumTokens()).resolves.toMatchSnapshot();
    expect(execa).toHaveBeenCalledTimes(1);
    expect(fs.readFile).toHaveBeenCalledTimes(1);
  });
});

describe('getAssets', () => {
  it('fetches all assets', async () => {
    await expect(getAssets()).resolves.toMatchSnapshot();
  });
});

describe('getParsedAssets', () => {
  it('fetches all parsed assets', async () => {
    await expect(getParsedAssets()).resolves.toMatchSnapshot();
  });

  it('removes invalid assets', async () => {
    const assets = await getParsedAssets();
    console.log(assets);
    expect(Object.keys(assets).length).toBe(2);
    expect(Object.values(assets)).not.toContainEqual({
      coinCapId: 'ethereum',
      coinGeckoId: 'ethereum',
      cryptoCompareId: 'ETH',
      cryptoCurrencyIconName: 'eth',
      dexAgId: 'ETH',
      invalid: 'field'
    });
  });
});

describe('writeToDisk', () => {
  it('writes an object to the disk', async () => {
    const assets = [
      {
        address: '0xa74476443119A942dE498590Fe1f2454d7D4aC0d',
        decimal: 18,
        name: 'Golem',
        symbol: 'GNT',
        uuid: '59a6f871-34bb-5914-86cc-b5f1307bb3b9'
      },
      {
        address: '0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2',
        decimal: 18,
        name: 'MakerDAO',
        symbol: 'MKR',
        uuid: 'd3230312-cacb-52e6-a849-8bbf6e8142dd'
      }
    ];

    await expect(writeToDisk(assets)).resolves.toBeUndefined();
    expect(fs.writeFile).toHaveBeenCalledTimes(1);
    expect(
      (fs.writeFile as jest.MockedFunction<typeof fs.writeFile>).mock.calls[0][1]
    ).toMatchSnapshot();
  });
});
