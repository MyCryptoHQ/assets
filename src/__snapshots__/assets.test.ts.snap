// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchEthereumTokens fetches Ethereum tokens from the disk 1`] = `
Array [
  Object {
    "address": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
    "decimal": 18,
    "name": "Golem",
    "symbol": "GNT",
    "uuid": "59a6f871-34bb-5914-86cc-b5f1307bb3b9",
  },
  Object {
    "address": "0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2",
    "decimal": 18,
    "name": "MakerDAO",
    "symbol": "MKR",
    "uuid": "d3230312-cacb-52e6-a849-8bbf6e8142dd",
  },
]
`;

exports[`getAssets fetches all assets 1`] = `
Array [
  Object {
    "name": "Ethereum",
    "symbol": "ETH",
    "uuid": "356a192b-7913-504c-9457-4d18c28d46e6",
  },
  Object {
    "name": "Ethereum Classic",
    "symbol": "ETC",
    "uuid": "6c1e671f-9af5-546d-9c1a-52067bdf0e53",
  },
  Object {
    "address": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
    "decimal": 18,
    "name": "Golem",
    "symbol": "GNT",
    "uuid": "59a6f871-34bb-5914-86cc-b5f1307bb3b9",
  },
  Object {
    "address": "0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2",
    "decimal": 18,
    "name": "MakerDAO",
    "symbol": "MKR",
    "uuid": "d3230312-cacb-52e6-a849-8bbf6e8142dd",
  },
]
`;

exports[`getIds gets all (available) service IDs for an asset 1`] = `
Object {
  "coinCapId": "golem-network-tokens",
  "coinGeckoId": "golem",
  "cryptoCompareId": "GNT",
  "dexAgId": "GNT",
}
`;

exports[`getParsedAssets fetches all parsed assets 1`] = `
Object {
  "59a6f871-34bb-5914-86cc-b5f1307bb3b9": Object {
    "coinCapId": "golem-network-tokens",
    "coinGeckoId": "golem",
    "cryptoCompareId": "GNT",
    "cryptoCurrencyIconName": "gnt",
    "dexAgId": "GNT",
  },
  "d3230312-cacb-52e6-a849-8bbf6e8142dd": Object {
    "coinCapId": "maker",
    "coinGeckoId": "maker",
    "cryptoCompareId": "MKR",
    "cryptoCurrencyIconName": "mkr",
    "dexAgId": "MKR",
  },
}
`;

exports[`parseTokensJson parses a JSON string to a tokens array 1`] = `
Array [
  Object {
    "address": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
    "decimal": 18,
    "name": "Golem",
    "symbol": "GNT",
    "uuid": "59a6f871-34bb-5914-86cc-b5f1307bb3b9",
  },
  Object {
    "address": "0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2",
    "decimal": 18,
    "name": "MakerDAO",
    "symbol": "MKR",
    "uuid": "d3230312-cacb-52e6-a849-8bbf6e8142dd",
  },
]
`;

exports[`writeToDisk writes an object to the disk 1`] = `
"[
  {
    \\"address\\": \\"0xa74476443119A942dE498590Fe1f2454d7D4aC0d\\",
    \\"decimal\\": 18,
    \\"name\\": \\"Golem\\",
    \\"symbol\\": \\"GNT\\",
    \\"uuid\\": \\"59a6f871-34bb-5914-86cc-b5f1307bb3b9\\"
  },
  {
    \\"address\\": \\"0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2\\",
    \\"decimal\\": 18,
    \\"name\\": \\"MakerDAO\\",
    \\"symbol\\": \\"MKR\\",
    \\"uuid\\": \\"d3230312-cacb-52e6-a849-8bbf6e8142dd\\"
  }
]"
`;
